File: student-geolocation-java/src/earth/ClassRoom.java
package earth;
import java.util.ArrayList;
import java.util.List;

public class ClassRoom {
	
	private String name;
	private Geolocation geolocation;
	private int dimention = 20;
	private List<Geolocation> squarePoints;
	
	public ClassRoom(String name, double latitude, double longitude) {
		this.name = name;
		geolocation = new Geolocation(latitude,longitude);
		squarePoints = new ArrayList<Geolocation>();
		fillSquerePoints();
	}
	
	/**
	 * Function to calculate the geolocation points (North-East, South-East, South-West,
	 * North-West) in base a geolocation point in the earth where is create a classroom 
	 * with dimension NxN. 
	 * Then the classroom know their points in the earth wich represent a square. 
	 */
	private void fillSquerePoints() {
		double meters = dimention/2;
		// 1 degree in google map is equal to 111.32 Kilometer. 1Degree = 111.32KM. 
		// 1KM in Degree = 1 / 111.32 = 0.008983. 1M in Degree = 0.000008983
		// 0.0000089 ~= coefficient of variation
		double coef = meters * 0.0000089; 
		double divLat = Math.cos(geolocation.getLatitude() * 0.018); // pi/180 ~= 0.018 

		Geolocation ne = new Geolocation(geolocation.getLatitude() + coef,
				                         geolocation.getLongitude() + coef/divLat);
		Geolocation se = new Geolocation(geolocation.getLatitude() - coef,
				                         geolocation.getLongitude() + coef/divLat);
		Geolocation sw = new Geolocation(geolocation.getLatitude() - coef,
				                         geolocation.getLongitude() - coef/divLat);
		Geolocation nw = new Geolocation(geolocation.getLatitude() + coef,
				                         geolocation.getLongitude() - coef/divLat);
		squarePoints.add(ne);
		squarePoints.add(se);
		squarePoints.add(sw);
		squarePoints.add(nw);
	}
	
	public String getName() {
		return name;
	}
	
	public void setName(String name) {
		this.name = name;
	}
	
	public Geolocation getGeolocation() {
		return geolocation;
	}
	
	public void setGeolocation(Geolocation geolocation) {
		this.geolocation = geolocation;
	}
	/**
	 * Algorithm using the 'haversine' formula to calculate the distance between two
	 * latitude/longitude points
	 * @see https://en.wikipedia.org/wiki/Haversine_formula
	 * @see http://www.movable-type.co.uk/scripts/latlong.html
	 * @param p1 Geolocation that contain latitude/longitude
	 * @param p2 Geolocation that contain latitude/longitude
	 * @return return the distance in meters
	 */
	public double calculatePoint(Geolocation p1, Geolocation p2) {
		int r = 6378137; // Earth’s mean radius in meter                                                                                           
		double dLat = rad(p2.getLatitude() - p1.getLatitude());                                                                                                           
		double dLong = rad(p2.getLongitude() - p1.getLongitude());                                                                                                          
	    double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) + 
	    		   Math.cos(rad(p1.getLatitude()) * Math.cos(rad(p2.getLatitude()))) *                                                                                    
			       Math.sin(dLong / 2) * Math.sin(dLong / 2);
		double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
		double d = r * c;
		return d; // returns the distance in meters                                                                                                   
	}
	
	/** 
	 * @param x represent a difference between latA - latB or longA - longB 
	 * @return a value in radians
	 */
	 private double rad (double x) {                                                                                                                        
		 return x * (Math.PI / 180);                                                                                                                  
	 };
	 
	 /**
	  * Function to verify if a geolocation point is inside of any polygon,
	  * This algorithm is bases point in polygon (PIP) used in computer graphics
	  * @see https://en.wikipedia.org/wiki/Point_in_polygon
	  * @see https://wrf.ecse.rpi.edu//Research/Short_Notes/pnpoly.html 
	  * @param geo is a object that has a latitude and longitude
	  * @return if the student is inside a course in base his geolocation 
	  */
	 public boolean isStudentInClassRoom (Geolocation geo) {
		 boolean inside = false;
		 double x = geo.getLatitude();
		 double y = geo.getLongitude();
		 for (int i = 0, j = squarePoints.size() - 1; i < squarePoints.size(); j = i++) {
		        double xi = squarePoints.get(i).getLatitude(); 
		        double yi = squarePoints.get(i).getLongitude();
		        double xj = squarePoints.get(j).getLatitude(); 
		        double yj = squarePoints.get(j).getLongitude();
		        
		        boolean intersect = ((yi > y) != (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
		        if (intersect) inside = !inside;
		}
		return inside;
	 }
	 
	 public void printSquarePoints() {
		 for(Geolocation g:squarePoints) {
			 System.out.println(g.getLatitude()+","+g.getLongitude());
		 }
	 }
	 
	 public void printDistances() {
		 for(Geolocation g:squarePoints) {
			 System.out.println(calculatePoint(geolocation,g));
		 }
	 }
}

File: student-geolocation-python/student_geolocation/models/classroom.py
import math

from geolocation import Geolocation


class ClassRoom:

    """ClassRoom class with name, latitude, and longitude as parameters"""

    def __init__(self, name, latitude, longitude):
        self.name = name
        self.geolocation = Geolocation(latitude, longitude)
        self.square_points = []
        self.dimention = 20
        self.fill_squere_points()

    @property
    def name(self):
        return self.name

    @property
    def dimention(self):
        return self.dimention

    @property
    def geolocation(self):
        return self.geolocation

    def fill_squere_points(self):
        """
        Function to calculate the geolocation points (North-East, South-East,
        South-West, North-West) in base a geolocation point in the earth where
        is create a classroom with dimension NxN.
        Then the classroom know their points in the earth wich represent a
        square.
        """
        meters = self.dimention/2
        # 1 degree in google map is equal to 111.32 Km. 1Degree = 111.32KM.
        # 1KM in Degree = 1 / 111.32 = 0.008983. 1M in Degree = 0.000008983
        # 0.0000089 ~= coefficient of variation
        coefficient_variation = 0.0000089
        coef = meters * coefficient_variation
        divLat = math.cos(math.radians(self.geolocation.latitude))
        ne = Geolocation(self.geolocation.latitude + coef,
                         self.geolocation.longitude + coef/divLat)
        se = Geolocation(self.geolocation.latitude - coef,
                         self.geolocation.longitude + coef/divLat)
        sw = Geolocation(self.geolocation.latitude - coef,
                         self.geolocation.longitude - coef/divLat)
        nw = Geolocation(self.geolocation.latitude + coef,
                         self.geolocation.longitude - coef/divLat)
        self.square_points.append(ne)
        self.square_points.append(se)
        self.square_points.append(sw)
        self.square_points.append(nw)

    def is_student_in_class_room(self, geo):
        """
        Function to verify if a geolocation point is inside of any polygon,
        This algorithm is bases point in polygon (PIP) used in computer
        graphics.
        @see https://en.wikipedia.org/wiki/Point_in_polygon
        @see https://wrf.ecse.rpi.edu//Research/Short_Notes/pnpoly.html

        @param geo: is a object that has a latitude and longitude
        @type geo: Golocation
        @rtype: boolean
        @return: if the student is inside a course in base his geolocation
        """
        inside = False
        x = geo.latitude
        y = geo.longitude
        size = len(self.square_points)
        i = 0
        j = size - 1
        while i < size:
            xi = self.square_points[i].latitude
            yi = self.square_points[i].longitude
            xj = self.square_points[j].latitude
            yj = self.square_points[j].longitude
            intersect = ((yi > y) !=
                         (yj > y)) and (x < (xj - xi) *
                                        (y - yi) / (yj - yi) + xi)
            if intersect:
                inside = True
            i += 1
            j = i

        return inside
